src/geom/SphPoint.cc

```diff
@@ -0,0 +1,244 @@
+        double const havDAlpha = sinLon * sinLon;
+        double const havD = havDDelta + cosLat1 * cosLat2 * havDAlpha;
+        double const sinDHalf = sqrt(havD);
+        return (2.0 * asin(sinDHalf)) * radians;
+    }
+
+    SphPoint::SphPoint(Angle const & longitude, Angle const & latitude) :
+                    _longitude(wrap(longitude).asRadians()),
+                    _latitude(latitude.asRadians()) {
+        if (!isfinite(_latitude) || !isfinite(_longitude)) {
+            throw pexExcept::InvalidParameterError("NaN detected in constructing (" +
+                                                   to_string(longitude.asDegrees()) +
+                                                   ", " + to_string(latitude.asDegrees()) +
+                                                   ")");
+        }
+        // For some reason abs(double) is ignored
+        if (fabs(_latitude) > HALFPI) {
+            throw pexExcept::OutOfRangeError("Angle " + to_string(latitude.asDegrees()) +
+                " is not a valid latitude.");
+        }
+    }
+
+    SphPoint::SphPoint(Point3D const & vector) {
+        double const norm = sqrt(vector.distanceSquared(Point3D()));
+        if (!isfinite(norm)) {
+            stringstream buffer;
+            buffer << "NaN detected in vector " << vector << ".";
+            throw pexExcept::InvalidParameterError(buffer.str());
+        }
+        if (norm <= 0.0) {
+            stringstream buffer;
+            buffer << "Vector " << vector << " has no direction.";
+            throw pexExcept::InvalidParameterError(buffer.str());
+        }
+
+        double const x = vector.getX() / norm;
+        double const y = vector.getY() / norm;
+        double const z = vector.getZ() / norm;
+
+        // Need roundabout approach to handle _longitude = -1e-16
+        _longitude = wrap(atan2(y, x) * radians).asRadians();
+        _latitude = asin(z);
+    }
+
+    Point3D SphPoint::getVector() const noexcept {
+        return Point3D(cos(_longitude)*cos(_latitude),
+                       sin(_longitude)*cos(_latitude),
+                                       sin(_latitude));
+    }
+
+    Angle SphPoint::operator[](size_t index) const {
+        switch (index) {
+          case 0:
+            return getLongitude();
+          case 1:
+            return getLatitude();
+          default:
+            throw pexExcept::OutOfRangeError("Index " + to_string(index) + " must be 0 or 1.");
+        }
+    }
+
+    bool SphPoint::operator==(SphPoint const & other) const noexcept {
+        if (abs(_latitude) < HALFPI) {
+            return _longitude == other._longitude && _latitude == other._latitude;
+        } else {
+            return _latitude == other._latitude;
+        }
+    }
+
+    bool SphPoint::operator!=(SphPoint const & other) const noexcept {
+        return !(*this == other);
+    }
+
+    Angle SphPoint::bearingTo(SphPoint const & other) const {
+        if (abs(_latitude) >= HALFPI) {
```

Comment:
Another instance of needing `std::abs` or `fabs` or `std::fabs`.

I am worried that your unit tests did not find this.

Also, I suggest you use the `atPole()` method here (should you agree to add it).